File: src/main/java/com/undsf/mcmodmgr/models/SemanticVersion.java
package com.undsf.mcmodmgr.models;

import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.NotNull;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * 语义化版本解析
 */
@Slf4j
public class SemanticVersion implements Comparable<SemanticVersion> {
    public static final Pattern SEMVER_PATTERN = Pattern.compile("(\\d+)(\\.\\d+)?(\\.\\d+)?(-[0-9A-Za-z]+)?");

    public Integer major;

    public Integer minor;

    public Integer patch;

    public String suffix;

    public SemanticVersion(String version) {
        parse(version);
    }

    public SemanticVersion(Integer major, Integer minor, Integer patch) {
        this.major = major;
        this.minor = minor;
        this.patch = patch;
    }

    public SemanticVersion(Integer major, Integer minor, Integer patch, String suffix) {
        this.major = major;
        this.minor = minor;
        this.patch = patch;
    }

    public int getMajor() {
        return major != null ? major : 0;
    }

    public int getMinor() {
        return minor != null ? minor : 0;
    }

    public int getPatch() {
        return patch != null ? patch : 0;
    }

    public String getSuffix() {
        return suffix != null ? suffix : "";
    }

    public boolean parse(String version) {
        Matcher matcher = SEMVER_PATTERN.matcher(version);
        if (matcher.find()) {
            String majorStr = matcher.group(1);
            String minorStr = matcher.group(2);
            String patchStr = matcher.group(3);

            try {
                major = Integer.parseInt(majorStr);

                if (minorStr != null && minorStr.length() > 1) {
                    minor = Integer.parseInt(minorStr.substring(1));
                }

                if (patchStr != null && patchStr.length() > 1) {
                    patch = Integer.parseInt(patchStr.substring(1));
                }
            }
            catch (NumberFormatException ex) {
                log.warn("版本号`{}`解析失败！", version, ex);
                return false;
            }
        }
        else {
            log.warn("不匹配");
        }
        return false;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == null) {
            return false;
        }

        if (obj instanceof SemanticVersion o) {
            if (getMajor() == o.getMajor() &&
                    getMinor() == o.getMinor() &&
                    getPatch() == o.getPatch() &&
                    getSuffix().equals(o.getSuffix())) {
                return true;
            }
        }

        return false;
    }

    @Override
    public int compareTo(@NotNull SemanticVersion o) {
        int deltaMajor = getMajor() - o.getMajor();
        if (deltaMajor != 0) return deltaMajor;

        int deltaMinor = getMinor() - o.getMinor();
        if (deltaMinor != 0) return deltaMinor;

        int deltaPatch = getPatch() - o.getPatch();
        if (deltaPatch != 0) return deltaPatch;

        return getSuffix().compareTo(o.getSuffix());
    }

    @Override
    public String toString() {
        StringBuilder version = new StringBuilder();

        if (major != null) {
            version.append(major);
        }

        if (minor != null) {
            version.append(".");
            version.append(minor);
        }

        if (patch != null) {
            version.append(".");
            version.append(patch);
        }

        if (suffix != null && !suffix.isEmpty()) {
            version.append("-");
            version.append(suffix);
        }

        return version.toString();
    }
}

File: src/main/kotlin/com/undsf/mcmodmgr/commons/SemanticVersion.kt
package com.undsf.mcmodmgr.commons

import mu.KotlinLogging

private val log = KotlinLogging.logger {}

/**
 * 语义化版本
 */
class SemanticVersion constructor(
        major: Int,
        minor: Int,
        patch: Int,
        preRelease: String,
        build: String
) : Comparable<SemanticVersion> {
    var major: Int
    var minor: Int
    var patch: Int
    var preRelease: String
    var build: String

    init {
        this.major = major
        this.minor = minor
        this.patch = patch
        this.preRelease = preRelease
        this.build = build
    }

    companion object {
        val VERSION_REGEX = Regex(pattern = "^(0|[1-9]\\d*)(\\.(0|[1-9]\\d*))?(\\.(0|[1-9]\\d*))?(?:-((?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\\+([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?\$")

        val MIN = SemanticVersion(0, 0, 0, "", "")
        val MAX = SemanticVersion(Int.MAX_VALUE, Int.MAX_VALUE, Int.MAX_VALUE, "", "")

        fun parse(verStr: String): SemanticVersion? {
            val result = VERSION_REGEX.find(verStr)
            if (result == null) {
                log.warn { "无法解析版本号：${verStr}" }
                return null
            }

            var major = if (result.groupValues[1].isNotEmpty()) result.groupValues[1].toInt() else 0
            var minor = if (result.groupValues[3].isNotEmpty()) result.groupValues[3].toInt() else 0
            var patch = if (result.groupValues[5].isNotEmpty()) result.groupValues[5].toInt() else 0
            var preRelease: String = result.groupValues[6]
            var build: String = result.groupValues[7]
            return SemanticVersion(major, minor, patch, preRelease, build)
        }
    }

    override fun equals(other: Any?): Boolean {
        if (other !is SemanticVersion) {
            return false
        }
        return compareTo(other) == 0
    }

    override fun compareTo(other: SemanticVersion): Int {
        var delta = major - other.major
        if (delta != 0) return delta

        delta = minor - other.minor
        if (delta != 0) return delta

        delta = patch - other.patch
        if (delta != 0) return delta

        delta = preRelease.compareTo(other.preRelease)
        if (delta != 0) return delta

        return build.compareTo(other.build)
    }

    override fun toString() : String {
        var version = "$major.$minor.$patch"
        if (preRelease != null && preRelease!!.isNotEmpty()) {
            version += "-$preRelease"
        }
        if (build != null && build!!.isNotEmpty()) {
            version += "-$build"
        }
        return version
    }
}